pybmix/core/pybmixcpp/python_hierarchy.cc

```diff
@@ -0,0 +1,233 @@
+#include "python_hierarchy.h"
+
+#include <google/protobuf/stubs/casts.h>
+#include <pybind11/embed.h>
+#include <pybind11/pybind11.h>
+
+#include <Eigen/Dense>
+#include <random>
+#include <sstream>
+#include <stan/math/prim/prob.hpp>
+#include <string>
+#include <vector>
+
+#include "algorithm_state.pb.h"
+#include "hierarchy_prior.pb.h"
+#include "ls_state.pb.h"
+#include "pybmixcpp/bayesmix/src/utils/rng.h"
+
+namespace py = pybind11;
+using namespace py::literals;
+
+extern py::module_ numpy;
+extern py::module_ fun;
+extern py::module_ numpy_random;
+extern py::object py_engine;
+extern py::object py_gen;
+
+void synchronize_cpp_to_py_state(const std::mt19937 &cpp_gen,
+                                 py::object &py_gen);
+
+void synchronize_py_to_cpp_state(std::mt19937 &cpp_gen,
+                                 const py::object &py_gen);
+
+py::list vector_to_list(const std::vector<double> &x);
+
+std::vector<double> list_to_vector(py::list &x);
+
+py::list eigen_to_list(const Eigen::RowVectorXd &x);
+
+//! PYTHON
+double PythonHierarchy::like_lpdf(const Eigen::RowVectorXd &datum) const {
+    py::list state_py = vector_to_list(state.generic_state);
+    py::list datum_py = eigen_to_list(datum);
+    double result = fun.attr("like_lpdf")(datum_py, state_py).cast<double>();
```

Comment:
Does it get faster if instead of calling `fun.attr("like_lpdf")` each time, we store

```?? like_lpdf_evaluator = fun.attr("like_lpdf")``` 

and so on for all of the involved functions?

Reply:
We'll provide the results with these changes soon, but we would like to highlight the fact that "fun" object is declared as a global variable already,

Reply:
Yes, I noticed that. What I'm saying is that we don't know what happens underneath the call `fun.attr("like_lpdf")` and it might be that the call is indeed expansive.

Since, in the end `fun.attr("like_lpdf")` returns something alike an `std::function` or a functor, it would be good practice to avoid calling `fun.attr("like_lpdf")` a gazillion times, when this function/functor could be stored inside as a member of the hierarchy itself

Reply:
We have tried to declare them as class members, but there is an issue we don't clearly get yet (something relating to the pybind11 python interpreter stuff, like described [here](https://github.com/pybind/pybind11/issues/1389)). As a workaround solution we tried to declare them as global variables (yeah, we know, it's a bad practice in general) and that resulted to 8% time reduction. 
Now we're studying more in details the object-oriented characteristics of pybind11 and will get back on this with a better implementation hopefully soon.

Reply:
No problem on having them global on my side, just declare them into a namespace. After all, these are proper functions, so I don't have moral issues with globals in this case

Reply:
Thank you. We have updated the code accordingly (namespace `py_global`).